### PR TITLE
Lifecycle Reporting Update

### PR DESCRIPTION
**Pull request by glevitzky**
Refactored lifecycle reporting logic out of A4A and into network impls. This was done by adding an overridable function to A4A that is called at various stages of the lifecycle, which is then overridden in the ad network impls to emit CSI pings.

I'm not happy with 'emitLifecycleEvent' as a name, so I'm open to suggestions for that and, of course, anything else.

### CODE REVIEW

**Comment by glevitzky** [comment]
FYI: The test currently failing in test-example-validation is not related to my code. I ran the test against master and it fails at the same place.

**Comment by glevitzky** [comment]
Yes, this is a short term solution, and the long term solution is still planned to involve amp-analytics.

File: extensions/amp-ad/0.1/amp-ad.js [hunk 1]
```diff
@@ -52,6 +52,9 @@ export class AmpAd extends AMP.BaseElement {
         // Unspecified or empty type.  Nothing to do here except bail out.
         return null;
       }
+      window.ampAdSlotIdCounter = window.ampAdSlotIdCounter || 0;
+      const slotId = window.ampAdSlotIdCounter++;
+      this.element.setAttribute('data-slot-id', slotId);
```
Comment (glevitzky) [inline extensions/amp-ad/0.1/amp-ad.js]: I don't believe so, but @keithwrightbos can correct me if I'm wrong.

Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad.js]: What about ad networks that have their own data-slot-id attribute? would that cause a conflict?

Comment (glevitzky) [inline extensions/amp-ad/0.1/amp-ad.js]: If I rename it to "data-amp-slot-id", would that cut down the risk of collision sufficiently?

Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad.js]: Yep, that definitely sounds better.  How about `data-amp-slot-index`? To not be confused with the other slot ID concept.

Comment (glevitzky) [inline extensions/amp-ad/0.1/amp-ad.js]: Sounds good; will do.

**Comment by lannka** [comment]
@glevitzky oopps conflict.

**Pull request merged**